File: backend-go/service/translation_service.go
package service

import (
	"context"

	"github.com/sargy/gridpulse/backend-go/model"
	"github.com/sargy/gridpulse/backend-go/repository"
)

type TranslationService struct {
	repo *repository.TranslationRepository
}

func NewTranslationService(repo *repository.TranslationRepository) *TranslationService {
	return &TranslationService{repo: repo}
}

// GetTranslations returns translation keys for the given language, falling
// back to English if the requested language is not found.
func (s *TranslationService) GetTranslations(ctx context.Context, lang string) (map[string]string, error) {
	t, err := s.repo.FindByLang(ctx, lang)
	if err == nil && t != nil {
		return t.Keys, nil
	}
	// Fallback to English
	t, err = s.repo.FindByLang(ctx, "en")
	if err != nil {
		return map[string]string{}, nil
	}
	return t.Keys, nil
}

func (s *TranslationService) Count(ctx context.Context) (int64, error) {
	return s.repo.Count(ctx)
}

func (s *TranslationService) Save(ctx context.Context, t model.Translation) error {
	return s.repo.InsertOne(ctx, t)
}

File: backend/src/main/java/com/gridpulse/api/service/TranslationService.java
package com.gridpulse.api.service;

import com.gridpulse.api.model.Translation;
import com.gridpulse.api.repository.TranslationRepository;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.Optional;

@Service
public class TranslationService {

    private final TranslationRepository translationRepository;

    public TranslationService(TranslationRepository translationRepository) {
        this.translationRepository = translationRepository;
    }

    public Map<String, String> getTranslations(String lang) {
        Optional<Translation> translation = translationRepository.findByLang(lang);
        if (translation.isPresent()) {
            return translation.get().getKeys();
        }
        // Fallback to English
        return translationRepository.findByLang("en")
                .map(Translation::getKeys)
                .orElse(Map.of());
    }

    public long count() {
        return translationRepository.count();
    }

    public void save(Translation translation) {
        translationRepository.save(translation);
    }
}
